src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -1179,6 +1264,209 @@ static void uct_ib_mlx5_devx_md_cleanup(uct_ib_md_t *ibmd)
+    UCT_IB_MLX5DV_SET(create_mkey_in, in, opcode,
+                      UCT_IB_MLX5_CMD_OP_CREATE_MKEY);
+    UCT_IB_MLX5DV_SET(create_mkey_in, in, translations_octword_actual_size, 1);
+    UCT_IB_MLX5DV_SET(create_mkey_in, in, mkey_umem_id, memh->umem->umem_id);
+    UCT_IB_MLX5DV_SET64(create_mkey_in, in, mkey_umem_offset, 0);
+    UCT_IB_MLX5DV_SET(mkc, mkc, access_mode_1_0,
+                      UCT_IB_MLX5_MKC_ACCESS_MODE_MTT);
+    UCT_IB_MLX5DV_SET(mkc, mkc, a, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, rw, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, rr, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, lw, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, lr, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, crossing_target_mkey, 1);
+    UCT_IB_MLX5DV_SET(mkc, mkc, qpn, 0xffffff);
+    UCT_IB_MLX5DV_SET(mkc, mkc, pd, dvpd.pdn);
+    UCT_IB_MLX5DV_SET(mkc, mkc, mkey_7_0, md->mkey_tag);
+    UCT_IB_MLX5DV_SET64(mkc, mkc, start_addr, (intptr_t)address);
+    UCT_IB_MLX5DV_SET64(mkc, mkc, len, length);
+
+    ucs_assert(memh->cross_mr == NULL);
+    memh->cross_mr = uct_ib_mlx5_devx_obj_create(md->super.dev.ibv_context, in,
+                                                 sizeof(in), out, sizeof(out),
+                                                 "MKEY",
+                                                 uct_md_reg_log_lvl(0));
+    if (memh->cross_mr == NULL) {
+        status = UCS_ERR_IO_ERROR;
+        goto err_umem_dereg;
+    }
+
+    ucs_assert(memh->super.exported_lkey == UCT_IB_INVALID_MKEY);
+    memh->super.exported_lkey = (UCT_IB_MLX5DV_GET(create_mkey_out, out,
+                                                   mkey_index) << 8) |
+                                md->mkey_tag;
+
+    UCT_IB_MLX5DV_SET(allow_other_vhca_access_in, ein, opcode,
+                      UCT_IB_MLX5_CMD_OP_ALLOW_OTHER_VHCA_ACCESS);
+    UCT_IB_MLX5DV_SET(allow_other_vhca_access_in, ein,
+                      object_type_to_be_accessed, UCT_IB_MLX5_OBJ_TYPE_MKEY);
+    UCT_IB_MLX5DV_SET(allow_other_vhca_access_in, ein,
+                      object_id_to_be_accessed,
+                      memh->super.exported_lkey >> 8);
+    access_key = UCT_IB_MLX5DV_ADDR_OF(allow_other_vhca_access_in, ein,
+                                       access_key);
+    ucs_strncpy_zero(access_key, uct_ib_mkey_token,
+                     UCT_IB_MLX5DV_FLD_SZ_BYTES(alias_context, access_key));
+
+    ret = mlx5dv_devx_general_cmd(md->super.dev.ibv_context, ein, sizeof(ein),
+                                  eout, sizeof(eout));
+    if (ret) {
+        uct_md_log_mem_reg_error(0, "mlx5dv_devx_general_cmd() failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err_cross_mr_destroy;
+    }
+
+    return UCS_OK;
+
+err_cross_mr_destroy:
+    mlx5dv_devx_obj_destroy(memh->cross_mr);
+    memh->cross_mr = NULL;
+err_umem_dereg:
+    mlx5dv_devx_umem_dereg(memh->umem);
+    memh->umem = NULL;
+err_out:
+    memh->super.exported_lkey = UCT_IB_INVALID_MKEY;
+    return status;
+}
+
+#else
```

Comment:
maybe the proprocessor if could be inside uct_ib_mlx5_devx_reg_exported_key function so would not need to write the function header twice